File: lang.py/byte_buffer2/node.py
#!/usr/bin/env python

from byte_buffer2 import *
from enum import Enum

class NodeType(Enum):
    File = 1
    Dir = 2
    SymLink = 3
    HardLink = 4

class Node:
    def __init__(self, name, stream, type=NodeType.File) -> None:
        self.name = name
        self.path = ""
        self.source, self.extents = stream 
        self.alloc_size = sum(map(lambda e: e.size, self.extents))
        self.size = self.alloc_size
        self.parent = None
        self.children = []
        self.type = type

    def is_file(self):
        return self.type == NodeType.File

    def is_dir(self):
        return self.type == NodeType.Dir

    def read_all(self):
        res = bytearray()
        file = self.source
        for extent in self.extents:
            file.seek(extent.offset)
            res.extend(file.read(extent.size))

        return bytes(res)

    def export_to(self, path):
        with open(path, 'wb') as file:
            for extent in self.extents:
                addr, size = extent
                source.seek(addr)
                b = source.read(size)
                file.write(b)

    def __str__(self) -> str:
        return f"name: {self.name}, size:{hex(self.size)}"
                        
if __name__ == "__main__":
    pass

File: ruby.mruby/md.analyzer/node.cpp
#include "node.h"

#include <iostream>
#include <filesystem>

using namespace std;
      namespace fs = filesystem;
////////////////////////////////////////////////////////////////////////////////
//
//
//
////////////////////////////////////////////////////////////////////////////////
node::node(string const& p)
  : _path(p)
{
  auto ph = std::filesystem::path(_path);

  _name = ph.filename().string();
  _size = 1234;
}

node::~node()
{
}

auto node::name() -> string
{
  return _name;
}

auto node::path() -> string
{
  return _path;
}

auto node::size() -> int64_t
{
  return _size;
}

auto node::parent() -> string
{
  return fs::path(_path).parent_path().string();
}

auto node::deleted() -> bool
{
  return false;
}

auto node::seek(int pos) -> void
{
}

auto node::save_to(string const& res) -> bool
{
  cout << "save_to: " << res << endl;

  return true;
}

auto node::read(uint32_t size) -> vector<uint8_t>
{
  vector<uint8_t> v;

  return v;
}

////////////////////////////////////////////////////////////////////////////////
//
//
//
////////////////////////////////////////////////////////////////////////////////
